src/peft/tuners/cpt/model.py

```diff
@@ -0,0 +1,207 @@
+        self.set_updated_tokens()
+
+    def forward(self, indices):
+        """
+        Computes the prompt embeddings and applies delta adjustments.
+
+        Args:
+            indices (torch.Tensor): Indices of the tokens to be embedded.
+
+        Returns:
+            torch.Tensor: Sum of prompt embeddings and delta embeddings.
+        """
+        with torch.no_grad():
+            prompt_embeddings = self.embedding(indices)
+
+        self.projection()  # Apply epsilon-based projection
+        delta_prompt_embeddings = self.delta_embedding(indices)
+
+        return prompt_embeddings + delta_prompt_embeddings
+
+    def set_updated_tokens(self):
+        """
+        Sets up a backward hook to selectively update token gradients based on the CPT token type mask.
+        """
+        if self.config.CPT_prompt_tuning_init == PromptTuningInit.TEXT:
+            tensor_ICL_mask = torch.Tensor(self.config.CPT_tokens_type_mask).long()
+            mask_input_template = torch.remainder(tensor_ICL_mask, 4) == 1
+            mask_input = torch.remainder(tensor_ICL_mask, 4) == 2
+            mask_output_template = torch.remainder(tensor_ICL_mask, 4) == 3
+            mask = mask_input_template | mask_input | mask_output_template
+            mask = mask.view(-1, 1)
+        elif self.config.CPT_prompt_tuning_init == PromptTuningInit.RANDOM:
+            mask = torch.ones((self.config.num_virtual_tokens, 1)).long()
+
+        def backward_hook(grad):
+            grad = grad * mask.to(grad.device)  # Apply mask to gradients
+            return grad
+
+        self.delta_embedding.weight.register_hook(backward_hook)
+
+    def get_epsilon(self):
+        if self.config.CPT_prompt_tuning_init == "TEXT":
+            CPT_tokens_type_mask = self.config.CPT_tokens_type_mask
+        else:
+            CPT_tokens_type_mask = [2] * self.config.num_virtual_tokens
+
+        MIN_VALUE = 1e-10
+
+        # Calculate normalized epsilon values for input, output, and format tokens
+        normalized_format_eps = (
+            self.config.opt_projection_format_epsilon
+            * torch.sqrt(
+                torch.Tensor([self.config.token_dim / 2048])
+            )
+        )
+        normalized_input_eps = self.config.opt_projection_epsilon * torch.sqrt(
+            torch.Tensor([self.config.token_dim / 2048])
+        )
+
+        epsilon = torch.ones_like(torch.Tensor(CPT_tokens_type_mask)).to(torch.float32) * MIN_VALUE
+        CPT_tokens_type_mask = torch.Tensor(CPT_tokens_type_mask).long()
+
+        epsilon[(CPT_tokens_type_mask > 0) & (torch.remainder(CPT_tokens_type_mask, 4) == 1)] = normalized_format_eps
+        epsilon[(CPT_tokens_type_mask > 0) & (torch.remainder(CPT_tokens_type_mask, 4) == 3)] = normalized_format_eps
+        epsilon[(CPT_tokens_type_mask > 0) & (torch.remainder(CPT_tokens_type_mask, 4) == 2)] = normalized_input_eps
+
+        return epsilon
+
+    def projection(self):
```

Comment:
I wonder if it would promote readability if we made this method `get_projection` and then it returns `new_embeddings_weights` instead of assigning it directly. Then we can assign on the caller site. WDYT?

Reply:
My suggestion is to rename this method to `get_projection`. Then, in the last line, instead of `self.delta_embedding.weight.data = new_embeddings_weights`, just return `new_embeddings_weights`. It is then on the caller side that the `delta_embeddings` are updated.